### ISSUE TITLE
Segment visibility in slice views can become out-of-sync with current visibility state

### ISSUE THREAD

**Issue by jamesobutler** [Slicer/Slicer#7206]
## Summary

It can be observed when toggling the segment visibility in the Segment Editor that it can become out-of-sync with what is shown in the slice views until a mouse move event in the Slice views triggers an update.

https://github.com/Slicer/Slicer/assets/15837524/c699e812-14e9-4ae9-8a7c-ffaba000b457

## Steps to reproduce

1. Load a volume and apply a segment that you can see in all the slice view orientations (axial, sagittal, coronal)
2. Toggle the segment visibility quickly
3. Observe that the visibility in the slice views gets out of sync with the eyeball state and only fixes itself upon mouse move in the slice views.
4. If you toggle the segment visibility slowly, the update event in all the slice views can be seen as being pretty slow. The segment might disappear in 2 views, but then a half second later disappear in the third view.

## Expected behavior

Toggling the segment visibility should always be in sync with what is displayed in the slice views. Otherwise you could incorrectly toggle the button again thinking your going to get in the correct state, but really you are just fighting a display render issue.

## Environment
- Slicer version: Slicer-5.4.0
- Operating system: Windows

**Comment by jcfr** [comment]
After downloading Slicer 5.4, 5.6 and Preview build, I wasn't able t reproduce the issue on either Linux or Windows.

@jamesobutler I suggest we review and discuss this during tomorrow (April 2nd) Developer's meeting.

**Comment by lassoan** [comment]
I experienced similar issues before: On some computers sometimes after various visualization option changes, not all views are updated until the mouse is moved. It might be some low-level VTK, Qt, graphics driver issue. It is not very noticeable, as moving the mouse automatically fixes the rendering, but it is still not nice. @jamesobutler do you think the problem you have reported here could be the same issue?

**Comment by jamesobutler** [comment]
Yes, @lassoan I think you are describing what I was seeing here. Where it is not unique to segment visibility but likely any other visualization updates that sometimes have a delayed update until the mouse moves over the slice view.

@jcfr, yes I can aim to be at the developer's meeting tomorrow to discuss. I downloaded the most recent Slicer preview and was able to replicate (on Windows) what I originally posted in this issue. The important part of replicating is to quickly toggle on/off segment visibility. If I toggle either on/off just once, I usually see it update it in one slice view first and then the others after a short delay. I think the update gets confused and out-of-sync if I toggle again before the other slice views have finished the original update.

**Comment by lassoan** [comment]
Let's test after Qt (and VTK) is updated.

**Comment by jamesobutler** [comment]
I have tried with latest Slicer preview as of today which includes latest VTK 9.5 where I can replicate this issue still when the app is using the Intel UHD graphics on my mobile workstation. When it is set to my NVIDIA T1200 graphics, then I do not observe the visibility issue.

<img width="629" height="924" alt="Image" src="https://github.com/user-attachments/assets/fd66bedf-434c-446d-a853-2982b2c799e0" />

**Comment by lassoan** [comment]
@jamesobutler please test with `SLICER_OPENGL_PROFILE` variable set to `core` or `no` profile.

You can also try if changing the maximum update rate from the default 60 to a lower or higher value has any effect (for example, `slicer.app.layoutManager().sliceWidget('Red').sliceView().maximumUpdateRate=25`). It is not very likely, but might be possible that somehow the delayed update (triggered by the RequestTimer) arrives at a wrong time and that's why has no effect. Or maybe the https://github.com/commontk/CTK/blob/master/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp class has some wrong logic about when a `scheduleRender` results in `d->RenderWindow->Render()` (it looks good, but there are quite a lot of special cases, depending on where a requests come from and when, so maybe there is a case that is not handled well). For example, perhaps https://github.com/commontk/CTK/blob/fb91447e0e15327341619cc0cda744724408ebc6/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp#L260 should be moved above https://github.com/commontk/CTK/blob/fb91447e0e15327341619cc0cda744724408ebc6/Libs/Visualization/VTK/Widgets/ctkVTKAbstractView.cpp#L250.

Do you have a developer machine where this issue can be easily reproduced? If yes, then the best would be to experiment there. Otherwise I can try to build on my computer and then test on an old surface laptop.

**Comment by lassoan** [comment]
After many hours of debugging, I think I've found the (almost embarrassingly simple) fix - https://github.com/lassoan/CTK/commit/51bea365347597ccd33b6f760bac3728acf0fdb7

@jamesobutler please test if it fixes the problem on your computer, too.

**Comment by jamesobutler** [comment]
@lassoan Thanks for the investigation! I set my app back to run with the graphics "Power Saving" option to force the use of the integrated graphics which is when I observed the error. I have tested locally with your CTK fix and can confirm that it fixes this rendering issue for me too 👏🏻 . When using the "High Performance" graphics option I wasn't seeing the issue previously and with this fix it all seems to remain good too. Therefore issuing the bug fix PR I think makes sense at this point.

**Comment by jamesobutler** [comment]
This issue will actually be closed once the CTK hash is updated for Slicer. @lassoan or @jcfr Do you want to issue that PR?

**Comment by jcfr** [comment]
Ditto. I already have a look branch ready to update CTK. It will be pushed shortly.
